File: python/patterns/facade.py
class Projector:
    def on(self) -> None:
        print("Projector on")

    def off(self) -> None:
        print("Projector off")


class SoundSystem:
    def on(self) -> None:
        print("Sound system on")

    def off(self) -> None:
        print("Sound system off")


class StreamingApp:
    def play(self, movie: str) -> None:
        print(f"Streaming '{movie}'")

    def stop(self) -> None:
        print("Stop streaming")


class MovieNightFacade:
    def __init__(self) -> None:
        self.projector = Projector()
        self.sound = SoundSystem()
        self.app = StreamingApp()

    def start_movie(self, movie: str) -> None:
        self.projector.on()
        self.sound.on()
        self.app.play(movie)

    def end_movie(self) -> None:
        self.app.stop()
        self.sound.off()
        self.projector.off()


def run() -> None:
    print("\nFacade")
    movie_night = MovieNightFacade()
    movie_night.start_movie("The Brave Little Toaster")
    movie_night.end_movie()

File: csharp/Patterns/Facade.cs
using System;

namespace DesignPatterns.Patterns;

public static class Facade
{
    private class Projector
    {
        public void On() => Console.WriteLine("Projector on");
        public void Off() => Console.WriteLine("Projector off");
    }

    private class SoundSystem
    {
        public void On() => Console.WriteLine("Sound system on");
        public void Off() => Console.WriteLine("Sound system off");
    }

    private class StreamingApp
    {
        public void Play(string movie) => Console.WriteLine($"Streaming '{movie}'");
        public void Stop() => Console.WriteLine("Stop streaming");
    }

    private class MovieNightFacade
    {
        private readonly Projector _projector = new();
        private readonly SoundSystem _sound = new();
        private readonly StreamingApp _app = new();

        public void StartMovie(string movie)
        {
            _projector.On();
            _sound.On();
            _app.Play(movie);
        }

        public void EndMovie()
        {
            _app.Stop();
            _sound.Off();
            _projector.Off();
        }
    }

    public static void Run()
    {
        Console.WriteLine("\nFacade");
        var movieNight = new MovieNightFacade();
        movieNight.StartMovie("The Brave Little Toaster");
        movieNight.EndMovie();
    }
}
